Make SHA256 compression runtime pluggable

furszy opened the pull request:
Tackling the long-standing request #702.

Right now we ship our own SHA256 implementation, a standard baseline version that does not take advantage of any hardware-optimized instruction, and it cannot be accessed by the embedding application - it is for internal usage only.

This means embedding applications often have to implement or include a different version for their use cases, wasting space on constrained environments, and in performance-sensitive setups it forces them to use a slower path than what the platform provides. Many projects already rely on tuned SHA-NI / ARMv8 / or other hardware-optimized code, so always using the baseline implementation we ship within the library is not ideal.

These changes allow users to supply their own SHA256 compression function at runtime, while preserving the existing default behavior for everyone else. This is primarily intended for environments where the available SHA256 implementation is detected dynamically and recompiling the library with a different implementation is not feasible (equivalent build-time functionality will come in a follow-up PR).

It introduces a new API:

```C89
secp256k1_context_set_sha256_transform_callback(ctx, fn_transform)
```

This function installs the optimized SHA256 compression into the `secp256k1_context`, which is then used by all internal computations. Important: The provided function is verified to be output-equivalent to the original one.

As a quick example, using this functionality in Bitcoin-Core will be very straightforward: https://github.com/furszy/bitcoin-core/commit/f68bef06d95a589859f98fc898dd80ab2e35eb39

real-or-random reviewed (commented):
Concept ACK, great to see a PR for this!

furszy commented on CMakeLists.txt:
```diff
@@ -285,6 +285,7 @@ message("  extrakeys ........................... ${SECP256K1_ENABLE_MODULE_EXTRA
 message("  schnorrsig .......................... ${SECP256K1_ENABLE_MODULE_SCHNORRSIG}")
 message("  musig ............................... ${SECP256K1_ENABLE_MODULE_MUSIG}")
 message("  ElligatorSwift ...................... ${SECP256K1_ENABLE_MODULE_ELLSWIFT}")
+message("  External SHA256 ..................... ${SECP256K1_EXTERNAL_SHA256_PATH-NONE}")
```
.

hebasto reviewed (commented):
Concept ACK.

furszy commented on include/secp256k1.h:
```diff
@@ -92,6 +93,7 @@ typedef struct secp256k1_ecdsa_signature {
  * the message, the algorithm, the key and the attempt.
  */
 typedef int (*secp256k1_nonce_function)(
+    const secp256k1_context *ctx,
```
> So we can keep the struct here and add the context arg only to our built-in nonce function. Our ECDSA signing function (and Schnorr signing, and ECDH, etc.) will need to special-case the built-in nonce function because it has a different function signature. Not elegant but it will do the job.

Done. Spent quite a bit of time thinking about this, mainly because the so-called "default" function (the one exposed in the public header [here](https://github.com/bitcoin-core/secp256k1/blob/8d445730ec79a9fa722eea59b6b697c3b9b1d501/include/secp256k1.h#L641)) is not the one that will be used when a custom sha256 transform is provided.., which somewhat defeats the purpose of calling it "default". 
Still, after a few rounds of thought and discussions with theStack and sipa, I think this is fine. It’s expected that a different function is used when the user provides their own sha transform.

furszy commented on src/hash.h:
```diff
@@ -10,13 +10,20 @@
-static void secp256k1_sha256_initialize(secp256k1_sha256 *hash);
+static void secp256k1_sha256_initialize(secp256k1_sha256 *hash, sha256_transform_callback fn_transform);
 static void secp256k1_sha256_write(secp256k1_sha256 *hash, const unsigned char *data, size_t size);
 static void secp256k1_sha256_finalize(secp256k1_sha256 *hash, unsigned char *out32);
```
> My suggestion is just passing the context object to every internal function that needs the SHA256 callback.

I did think about that initially, but didn’t really like it. Mainly because going down that path will be significantly more invasive than the current changes: It requires changing every `secp256k1_sha256_write` call to receive the function arg, every hmac, tagged hash etc.
It also includes changing `secp256k1_sha256_finalize`! which will also require the function arg because we call write during finalization too.
Overall, I think the end result would be quite ugly for what we gain here (see below).

> A potential problem with storing the callback in the `secp256k1_sha256` struct is that you could have this series of events:
> 
> * User sets callback to `cb1`.
> * User calls API function that calls `secp256k1_sha256_initialize`, which outputs some opaque object that contains the `secp256k1_sha256` struct.
> * User sets callback go `cb2`.
> * Users passes the opaque object to another API function, which calls `secp256k1_sha256_write`, which calls `cb1`. UB.
> 
> (I believe) this cannot happen in the current code because there's no pair of API functions that behaves in this way, but it's a potential footgun for the future.

I'm not fully convinced it's a realistic scenario. Changing the sha256 transform callback in the middle of an ongoing procedure doesn’t really make sense in any intended usage. If someone does that, they should already be well outside the model we expect the API to be used under.
To me, this is similar to deleting the callback function from another thread while a libsecp operation is executing.

Also, exposing an in-use sha256 object across API boundaries is something we should simply never do right?. That feels really ugly. These structs are only valid under the assumption that they were initialized via `secp256k1_sha256_initialize`; otherwise, the internal fields would already be invalid, independently of the callback pointer. And we cannot ensure that the struct has been initialized properly (or not corrupted) if we share it externally.
I mean, if we ever want to share an existing sha256, I would say we should share the state and buffer directly as bytes and then re-initialize the struct on the receiving side.

So overall, I think it worth documenting the expectation clearly for sure: like sha256 structs must not be shared externally, and the provided sha256 callback, must be provided during app's init, and must remain active for the entire app lifecycle once provided (because it really doesn't make sense to change it). But I don’t think this is a problem we need to design around.

furszy replied:
Maybe I should give the "pass the function to every `write` call" patch-set another try and compare the resulting changes. I did it early on so maybe it is not as bad as I remember it. Let me see..

real-or-random replied:
> Also, exposing an in-use sha256 object across API boundaries is something we should simply never do right?. That feels really ugly.

It doesn't feel ugly to me. Of course, we wouldn't share it standalone but inside some other object. For example, like we expose an `secp256k1_scalar` inside an opaque `secp256k1_musig_keyagg_cache`. 

And I don't think the use case would be unrealistic. Having a "hash builder" is common, and it could very well be part of a high-level function. Just imagine an advanced schnorrsig signing object that lets you append to the message multiple times before finalizing the signatures (e.g., for cases in which you want to sign a long message whose parts you receive step by step).

furszy commented on src/secp256k1.c:
```diff
@@ -56,20 +57,27 @@
 
+struct secp256k1_hash_context {
+    sha256_transform_callback fn_sha256_transform;
+};
```
> I think it will be nice to define this in the internal hash module, just like secp256k1_ecmult_gen_context is defined in the ecmult_gen module.

Sure. Done as suggested.

furszy reviewed (commented):
Update pushed, thanks for the deep review real_or_random!
I haven’t fully finished addressing all the comments yet, but I'll hopefully finish tomorrow. So many good topics.

furszy commented on src/hash.h:
```diff
@@ -10,13 +10,20 @@
+
+/* Validate user-supplied SHA-256 transform by comparing its output against
+ * the library's linked implementation */
+static int secp256k1_sha256_check_transform(sha256_transform_callback fn_transform);
```
I missed to update this one. The latest push uses only `secp256k1_sha256_selftest`. Calling it twice.

furszy committed:
Introduce hash context to support pluggable SHA256 compression

This is purely a mechanical change with no behavior change.

It introduces a secp256k1_hash_ctx struct inside secp256k1_context
and propagates it to all SHA256-related operations.

This sets up the ability to provide a hardware-optimized SHA256
compression function at runtime in a follow-up commit.

furszy commented on configure.ac:
```diff
@@ -134,6 +134,12 @@ SECP_TRY_APPEND_DEFAULT_CFLAGS(SECP_CFLAGS)
+  AS_HELP_STRING([--with-external-sha256=PATH], [use external SHA256 compression implementation]),
+  [external_sha256_path="$withval"],
+  [external_sha256_path=""]
+)
```
Hmm, I see. Those are very good points. Agree this worth a deeper discussion.

No problem splitting the changes. We can keep this PR focused on the runtime feature and handle the build-time override design separately. The only change we need from the build-time commit for this PR is exposing the transform function in `hash.h`, so we can initialize the context default function.

real-or-random replied:
> We can keep this PR focused on the runtime feature and handle the build-time override design separately

Right, I hadn't considered this option. That's a good idea. (And I see that sipa suggested the same below.)

sipa commented:
A few high-level comments:
* I don't think we should do link-time plugging in the same PR as runtime plugging, and when we do, I don't think it should be done through the "module" abstractions. We use modules for exporting functionality from the library, not for offering replacement of internal dependencies (I find it strange to have base functionality depend on a module!).
* "round" is a term of art in cipher design, and SHA-256 has exactly 64 rounds. Using the same term with another meaning is confusing.
* I would try hard to avoid storing the transformation function pointer inside the `secp256k1_sha256` struct. Reviewing many lines that just add an extra parameter being pass through is not hard.

real-or-random commented on include/secp256k1.h:
```diff
@@ -403,6 +405,29 @@ SECP256K1_API void secp256k1_context_set_error_callback(
+ */
+SECP256K1_API void secp256k1_context_set_sha256_transform_callback(
+        secp256k1_context *ctx,
+        void (*sha256_transform_callback)(uint32_t *state, const unsigned char *block, size_t rounds)
```
> But this seems to be a low-level primitive file that shouldn't receive the context.

Hm, I don't think there is (or should be) a rule that says low-level primitives shouldn't receive the error callback. It doesn't need to be the full context.

But let me just resolve this conversation. As I already wrote above, I doubt that my suggestion was useful in the first place.

furszy reviewed (commented):
Per feedback, this is how it would look if we pass the function everywhere https://github.com/furszy/secp256k1/commit/8149fc119071c532e9d2d6c3b633af7a37e5a0b8 vs how it looks when the function lives in the `sha256` struct (an updated version of what we currently have here): https://github.com/furszy/secp256k1/commit/eb570826371f42a5bd320f8e43800281b1f63b0a

Let me know what you guys think.

sipa commented:
@furszy From a quick glance, the passing of the function everywhere is fine I think.

I'd suggest passing a pointer to `secp256k1_hash_context` to the sha256 functions rather than the function pointer itself. That makes it easier to (if ever) add additional fields to that context.

furszy reviewed (commented):
Updated per feedback. We now pass the hash context to all functions that need it, without storing any state inside the `sha256` struct. The API arg "rounds" has also been renamed to "blocks" to prevent confusion with existing terms. Also, the commit introducing the compile-time feature has been removed for inclusion at a later stage (see https://github.com/bitcoin-core/secp256k1/pull/1777#discussion_r2598064155 for more details).

real-or-random commented on src/secp256k1.c:
```diff
@@ -220,6 +229,18 @@ void secp256k1_context_set_error_callback(secp256k1_context* ctx, void (*fun)(co
     ctx->error_callback.data = data;
 }
 
+void secp256k1_context_set_sha256_transform_callback(secp256k1_context *ctx, void (*sha256_transform_callback_)(uint32_t *state, const unsigned char *block, size_t rounds)) {
```
If I'm not mistaken, this comment is still unaddressed.

Let me add that the typedef is a good place to document the behavior of the function (e.g., the `blocks` arg). The naming of the args should be reconsidered because now `block` may point to more than one block. Maybe `(..., const unsigned char *blocks64, size_t n_blocks)`.

furszy replied:
> If I'm not mistaken, this comment is still unaddressed.
> 
> Let me add that the typedef is a good place to document the behavior of the function (e.g., the `blocks` arg). The naming of the args should be reconsidered because now `block` may point to more than one block. Maybe `(..., const unsigned char *blocks64, size_t n_blocks)`.

Done, pushed. Sorry for missing this one. Most of the function docs were inside the build-time feature commit. Inside the module definition. I just moved what had here and renamed the args.

real-or-random commented:
> Nice, I hope I'll find the time have another look soon

Sorry, I didn't manage to find the time before the holidays...

real-or-random commented on include/secp256k1.h:
```diff
@@ -404,6 +404,49 @@ SECP256K1_API void secp256k1_context_set_error_callback(
 ) SECP256K1_ARG_NONNULL(1);
 
+/**
+ * SHA256 block compression raw primitive.
```
```suggestion
 * A pointer to an implementation of SHA256's internal compression function
```

a bit more in line with the other function pointer types

real-or-random reviewed (commented):
Approach ACK

furszy commented on include/secp256k1.h:
```diff
@@ -404,6 +404,49 @@ SECP256K1_API void secp256k1_context_set_error_callback(
+ *
+ * This function performs the SHA256 compression on one or more contiguous
+ * 64-byte message blocks. It does **not** perform padding, message scheduling
+ * across blocks, or length encoding.
```
> What is "message scheduling across blocks"?

Tried to state that the function operates only on complete 64-byte blocks and does not retain or combine leftover data between calls. E.g. calling it twice with two 32-byte inputs will not buffer them into a single 64-byte block; the function expects fully formed blocks and anything else is caller misuse.
Probably just saying that the function only accepts 64-byte **finalized** blocks is enough and better.

> "length encoding" is part of padding, no?

Yeah, I tend to think of padding as just the zero fill.., but specs wise this is wrong. The msg length is officially part of the padding.

furszy commented on src/testutil.h:
```diff
@@ -11,6 +11,14 @@
+    static void name(uint32_t *s, const unsigned char *msg, size_t rounds) {    \
+        name##_called = 1;                                                      \
+        secp256k1_sha256_transform(s, msg, rounds);                             \
+    }
```
what you mean with "custom hash function (not custom SHA256 transform)"?

The idea of the test was to verify that:
1) Default behavior remains unchanged when the ctx compression function is not set.
2) Custom compression function is actually called when provided. And because such function is correct, the complete ECDH test passes.

furszy commented on src/modules/ellswift/main_impl.h:
```diff
@@ -572,8 +580,15 @@ int secp256k1_ellswift_xdh(const secp256k1_context *ctx, unsigned char *output,
-    /* Invoke hasher */
-    ret = hashfp(output, sx, ell_a64, ell_b64, data);
+    /* Invoke hasher.
+     * Use context-aware function when custom SHA256 is provided */
```
Done as suggested.

sipa commented on include/secp256k1.h:
```diff
@@ -403,6 +404,47 @@ SECP256K1_API void secp256k1_context_set_error_callback(
+ * updates the internal SHA256 state accordingly. The function is not responsible
+ * for counting consumed blocks or bytes, nor for performing padding.
+ *
+ * Out:     state:     pointer to eight 32-bit words representing the current internal state;
```
This should probably be labelled "In/Out".

sipa commented on include/secp256k1.h:
```diff
@@ -403,6 +404,47 @@ SECP256K1_API void secp256k1_context_set_error_callback(
+ *
+ * Out:     state:     pointer to eight 32-bit words representing the current internal state;
+ *                     the state is updated in place.
+ * In:      blocks64:  pointer to the first 64-byte block; additional blocks must
```
"Pointer to concatenation of n_blocks blocks, of 64 bytes each". Maybe also worth mentioning that there is no alignment guarantee on the input?

sipa commented on include/secp256k1.h:
```diff
@@ -403,6 +404,47 @@ SECP256K1_API void secp256k1_context_set_error_callback(
+ * This API exists to support environments that wish to route the
+ * SHA256 compression step through a hardware-accelerated or otherwise
+ * specialized implementation. It is not meant for modifying the
+ * semantics of SHA256.
```
Maybe "It is not meant for replacing the use of SHA-256 in this library with a modified hash function"?

Modifying the semantics of SHA-256 sounds hard, as it may involve $5 wrench attacks on a few NIST officials.

furszy replied:
hehe sure, thanks.

furszy committed:
sha256: speed up writes using multi-block compression

Multiple 64-byte blocks can now be compressed directly
from the input buffer, without copying them into the
internal buffer.

real-or-random commented on src/hash_impl.h:
```diff
@@ -136,22 +136,34 @@ static void secp256k1_hash_context_init(secp256k1_hash_context *hash_ctx) {
         bufsize = 0;
     }
+
+    /* If we still have data to process; Invoke compression directly on the input */
```
nit: "If we still have data to process, invoke compression directly on the input" (why a semi-colon?)

furszy replied:
Done as suggested. Dunno how the semicolon and the capitalized "invoke" got there..
I probably had two different sentences and merged them at the end in the worst possible way.

real-or-random reviewed (commented):
two nits on the last commit (I like that one!)

furszy commented on src/modules/schnorrsig/main_impl.h:
```diff
@@ -49,7 +49,7 @@ static const unsigned char bip340_algo[] = {'B', 'I', 'P', '0', '3', '4', '0', '
 static const unsigned char schnorrsig_extraparams_magic[4] = SECP256K1_SCHNORRSIG_EXTRAPARAMS_MAGIC;
 
-static int nonce_function_bip340(unsigned char *nonce32, const unsigned char *msg, size_t msglen, const unsigned char *key32, const unsigned char *xonly_pk32, const unsigned char *algo, size_t algolen, void *data) {
+static int nonce_function_bip340_impl(const secp256k1_context *ctx, unsigned char *nonce32, const unsigned char *msg, size_t msglen, const unsigned char *key32, const unsigned char *xonly_pk32, const unsigned char *algo, size_t algolen, void *data) {
```
All suggestions applied (except the one above). Even when it was a mechanical change, took me more than what I expected to tackle them 🐜

real-or-random replied:
> All suggestions applied (except the one above). Even when it was a mechanical change, took me more than what I expected to tackle them 🐜

thanks for your stamina!

furszy committed:
Add API to override SHA256 compression at runtime

This introduces `secp256k1_context_set_sha256_compression()`,
which allows users to provide their own SHA256 block-compression
function at runtime.

This is useful in setups where the fastest implementation can only
be determined dynamically based on the available CPU features, and
rebuilding the library is not possible.

The callback is installed on the `secp256k1_context` and is then used
by all operations that compute SHA256 hashes. As part of the setup,
the library performs sanity checks to ensure that the supplied
function is equivalent to the default transform.

Passing NULL to the callback setter restores the built-in
implementation.

theStack reviewed (commented):
Concept ACK

I extended your example usage commit in Bitcoin Core to use the SHA256 HW acceleration also for signature verification functions (currently they wouldn't used it since we pass the static context) and based in on the new P2TR schnorr verification benchmark PR https://github.com/bitcoin/bitcoin/pull/34472: https://github.com/theStack/bitcoin/commit/296ec21c0e9b3612f07295e54b7ca2b2c1b10a19

The relevant benchmarks can be executed via:
```
$ ./build/bin/bench_bitcoin -filter=SignTransaction.*\|VerifyScript.*\|ConnectBlock.*
```

Unfortunately the benchmarks results fluctuate too much on the machine I'm currently working on, so posting the results wouldn't have much meaning yet. Will try to establish a stable environment and come back to that later. Maybe it is useful for other reviewers already.

sipa reviewed (commented):
ACK d6d14d84d07148d91479bdb4c1cf41a8d7231f9a

I agree with @real-or-random's comments above, though they seem minor. I'm happy to do a quick re-review if you should choose to address any of them.

real-or-random reviewed (approved):
ACK 4d92a083bc267b87db654ae75cbcaf2a87854c09

w0xlt reviewed (approved):
ACK 4d92a083bc267b87db654ae75cbcaf2a87854c09

theStack reviewed (approved):
ACK 4d92a083bc267b87db654ae75cbcaf2a87854c09

nit: I also tend to prefer direct `&ctx->hash_ctx` usage over the `secp256k1_get_hash_context` getter function, as suggested in https://github.com/bitcoin-core/secp256k1/pull/1777#issuecomment-3984558075 above.

sipa reviewed (commented):
Posthumous reACK 4d92a083bc267b87db654ae75cbcaf2a87854c09
